### PR TITLE
IBX-8117: Prepared Rector package

### PR DESCRIPTION
**Pull request by alongosz**
| :ticket: Issue | IBX-8117 |
|----------------|-----------|

#### Description:
This PR provides some skeleton which adapts https://github.com/rectorphp/rector to our needs.

Changes:
- [x] required `rector/rector` v1
- [x] forced `nikic/php-parser` `^4` due to rectorphp/rector#8567.
- [x] dropped integration tests setup as they seemed not to apply to this package needs (there's Rector-given test framework)
- [x] added placeholder for Ibexa DXP v5.0 ruleset
- [x] documented installation, usage, and writing custom rules guide
- [x] bumped PHPUnit to ^10 to be able to use Rector-provided PHPUnit template for testing custom rules
- [x] added test coverage for essential parts
- [x] defined a Composer-wrapped Symfony command `define-custom-rule` and necessary processors, based on built-in Rector `custom-rule` command, utilizing rector-provided templates

Background: built-in command is not very flexible and does too much, adding our custom command based on the built-in seemed like the most sane approach (though required not small portion of code). Road explored: wrap it and move files afterwards, but that doesn't provide the best DX as the command messes with composer and PHPUnit configuration...

Please note that this is executed by `composer define-custom-rule` so there's no Symfony DI involved there, nor bundle requirement. It makes it more compact and easier to install and use, though it's something that can be changed, depending on the preferences.

#### For QA:

Rather no QA at this point.

#### Documentation:
All needed information were documented here. Due to the nature of this package IMHO it makes more sense to do it here, however we can still either reference this in the Official Doc or move it and reference Official Doc here.

### CODE REVIEW

File: tests/lib/Composer/Command/CustomRule/RectorTemplateContentsProcessorTest.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            PHP,
+            <<<PHP
+            namespace Ibexa\Rector\Tests\Rule\Foo\Bar;
+
+            final class BazRector extends AbstractRector
+            PHP,
+        ];
+    }
+
+    protected function createSymfonySplFileInfoMock(string $fileContents): SplFileInfo
```
Comment (adamwojs) [inline tests/lib/Composer/Command/CustomRule/RectorTemplateContentsProcessorTest.php:87]:
```suggestion
    private function createSymfonySplFileInfoMock(string $fileContents): SplFileInfo
```

**Review by adamwojs — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -1,12 +1,46 @@
...
+
+## Installation
+
+```
+composer require --dev ibexa/rector:~5.0.x-dev
+```
+
+## Usage
+
+1. Create `./rector.php` file in your project, with the following contents, adjusted to your project structure:
```
Comment (mnocon) [inline README.md:19]: I suggest adding a recipe so that the file is created when the package is installed (this will make life easier for our partners)

**Review by mnocon — APPROVED** [review]
Looks good to me 💪

File: tests/lib/Composer/Command/CustomRule/RectorTemplateContentsProcessorTest.php (hunk 1 above)
Comment (alongosz) [reply tests/lib/Composer/Command/CustomRule/RectorTemplateContentsProcessorTest.php]: Nice catch, thanks.

File: src/lib/Composer/Command/CustomRule/AbstractRectorTemplateProcessor.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * @internal
+ */
+abstract class AbstractRectorTemplateProcessor
+{
+    protected function buildRectorName(string $ruleName): string
+    {
+        $rectorName = $ruleName;
+        // suffix with Rector by convention
+        if (substr_compare($rectorName, 'Rector', -strlen('Rector')) !== 0) {
```
Comment (Steveb-p) [inline src/lib/Composer/Command/CustomRule/AbstractRectorTemplateProcessor.php:20]: use `str_ends_with` instead.

```suggestion
        if (!str_ends_with($rectorName, 'Rector')) {
```

File: phpunit.xml.dist [hunk 4]
```diff
@@ -1,12 +1,7 @@
 <phpunit
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd"
```
[the remark this answers is not in the document]
Comment (alongosz) [reply phpunit.xml.dist]: Whoops, good catch

**Review by Steveb-p — APPROVED** [review]

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+* `./src/contracts/Sets` - directory to add global rulesets
+  - `./src/contracts/Sets/ibexa-50.php` is a ruleset which contains all the rules for Ibexa DXP 5.0 upgrade, new rules 
+    should be added there.
+* `./src/lib/Rule` - directory to implement specific rules, following [Rector's official doc](https://getrector.com/documentation/custom-rule).
+  - `./src/lib/Rule/Ibexa50/` - directory to implement Ibexa 5.0 upgrade specific rules
+* `./tests/lib/Rule` - directory to implement a specific rule tests, should contain subdirectory named `<RuleName>/`
+
+## Creating custom rules
+
+This package comes with a Composer-wrapped Symfony command to generate custom rule and place its files to proper
```
Comment (konradoboza) [inline CONTRIBUTING.md:17]:
```suggestion
This package comes with a Composer-wrapped Symfony command to generate custom rule and place its files in proper
```

**Review by konradoboza — APPROVED** [review]

File: CONTRIBUTING.md (hunk 5 above)
Comment (alongosz) [reply CONTRIBUTING.md:17]: Sorry, missed this. Fixed via ee2b174 on `main`, thanks.

**Pull request merged**